finality-aleph/src/validator_network/io.rs

```diff
@@ -0,0 +1,176 @@
+use std::fmt::{Display, Error as FmtError, Formatter};
+
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+
+use crate::validator_network::Data;
+
+// We allow sending up to 16MiB, that should be enough forever.
+pub const MAX_DATA_SIZE: u32 = 16 * 1024 * 1024;
+
+/// A general error when sending or receving data.
+#[derive(Debug, PartialEq, Eq)]
+pub enum Error {
+    ConnectionClosed,
+    DataTooLong(u32),
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        use Error::*;
+        match self {
+            ConnectionClosed => write!(f, "connection unexpectedly closed"),
+            DataTooLong(length) => write!(
+                f,
+                "encoded data too long - {} bytes, the limit is {}",
+                length, MAX_DATA_SIZE
+            ),
+        }
+    }
+}
+
+/// An error when sending data.
+#[derive(Debug, PartialEq, Eq)]
+pub struct SendError(Error);
+
+impl Display for SendError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl From<Error> for SendError {
+    fn from(e: Error) -> Self {
+        SendError(e)
+    }
+}
+
+/// An error when receiving data.
+#[derive(Debug, PartialEq, Eq)]
+pub enum ReceiveError {
+    Error(Error),
+    DataCorrupted,
+}
+
+impl Display for ReceiveError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        use ReceiveError::*;
+        match self {
+            Error(e) => write!(f, "{}", e),
+            DataCorrupted => write!(f, "received corrupted data"),
+        }
+    }
+}
+
+impl From<Error> for ReceiveError {
+    fn from(e: Error) -> Self {
+        ReceiveError::Error(e)
+    }
+}
+
+/// Sends some data using the stream.
+pub async fn send_data<S: AsyncWriteExt + Unpin, D: Data>(
+    mut stream: S,
+    data: D,
+) -> Result<S, SendError> {
+    let mut encoded = data.encode();
+    let len = u32::try_from(encoded.len()).map_err(|_| Error::DataTooLong(u32::MAX))?;
+    if len > MAX_DATA_SIZE {
+        return Err(Error::DataTooLong(len).into());
+    }
+    let mut payload = len.to_le_bytes().to_vec();
+    payload.append(&mut encoded);
+    stream
+        .write_all(&payload)
+        .await
+        .map_err(|_| Error::ConnectionClosed)?;
+    Ok(stream)
+}
+
+/// Attempts to receive some data using the stream.
+pub async fn receive_data<S: AsyncReadExt + Unpin, D: Data>(
+    mut stream: S,
+) -> Result<(S, D), ReceiveError> {
+    let mut buf = [0; 4];
+    stream
+        .read_exact(&mut buf[..])
+        .await
+        .map_err(|_| Error::ConnectionClosed)?;
+    let len = u32::from_le_bytes(buf);
+    if len > MAX_DATA_SIZE {
+        return Err(Error::DataTooLong(len).into());
+    }
+    let mut buf: Vec<u8> = vec![0; len as usize];
+    stream
+        .read_exact(&mut buf[..])
+        .await
+        .map_err(|_| Error::ConnectionClosed)?;
```

Comment:
What do you mean?

Reply:
It looks like it masks all possible network errors by mapping them to `ConnectionClosed`. It looks like one of first spots for debugging in case there are some connectivity problems.

Reply:
Welp, good point, for some reason I thought the underlying errors would be generic to `io`, but `io::Error` allows for custom errors, so they are probably informative. Changed. Annoyingly now we do not have `Eq` anymore, but that's a small price to pay.